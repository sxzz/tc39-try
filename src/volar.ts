import { Parser } from 'acorn'
import { tsPlugin } from 'acorn-typescript'
import { walk } from 'estree-walker'
import { createPlugin, type PluginReturn } from 'ts-macro'
import { tryOperatorPlugin } from './acorn'

const plugin: PluginReturn<undefined, false> = createPlugin(() => {
  // @ts-expect-error
  const parser = Parser.extend(tsPlugin()).extend(tryOperatorPlugin())

  return {
    name: 'tc39-try',
    resolveVirtualCode({ codes, ast }) {
      if (!ast.text.includes('try ')) return

      const program = parser.parse(ast.text, {
        ecmaVersion: 'latest',
        sourceType: 'module',
        allowAwaitOutsideFunction: true,
        allowHashBang: true,
        allowImportExportEverywhere: true,
        allowReturnOutsideFunction: true,
        allowSuperOutsideMethod: true,
      })

      let changed = false
      walk(program as any, {
        enter(node: any) {
          if (node.type === 'UnaryExpression' && node.operator === 'try') {
            let argumentStart = node.argument.start

            const isAwait = node.argument.type === 'AwaitExpression'
            if (isAwait) {
              argumentStart = node.argument.argument.start
            }

            codes.replaceRange(
              node.start,
              argumentStart,
              `${isAwait ? 'await ' : ''}__t(() => (`,
            )
            codes.replaceRange(node.argument.end, node.end, '))')
            changed = true
          }
        },
      })

      if (changed) {
        codes.push(`import { __t } from 'tc39-try';\n`)
      }
    },
  }
})

export default plugin
