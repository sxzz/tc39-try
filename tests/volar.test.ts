import path from 'node:path'
import { testFixtures } from '@sxzz/test-utils'
import { getLanguagePlugins } from '@ts-macro/language-plugin'
import { proxyCreateProgram } from '@volar/typescript'
import fg from 'fast-glob'
import * as ts from 'typescript'
import { describe } from 'vitest'
import volar from '../src/volar'

const workspace = path.resolve(__dirname, './fixtures')

describe('volar', async () => {
  const compilerOptions: ts.CompilerOptions = {
    rootDir: workspace,
    outDir: workspace,
    target: ts.ScriptTarget.ES2020,
    module: ts.ModuleKind.ESNext,
    allowNonTsExtensions: true,
    skipLibCheck: true,
  }
  const host = ts.createCompilerHost(compilerOptions)
  const createProgram = proxyCreateProgram(ts, ts.createProgram, (
    ts,
    options,
  ) => {
    return getLanguagePlugins(ts, options.options, { plugins: [volar()] })
  })
  const program = createProgram({
    options: compilerOptions,
    host,
    rootNames: await fg(`${workspace}/*.ts`),
  })

  await testFixtures(
    '*.ts',
    (_, id) => {
      const ast = program.getSourceFile(id)
      return ast?.text.trim()
    },
    { cwd: workspace },
  )
})
