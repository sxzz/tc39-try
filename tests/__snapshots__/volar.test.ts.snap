// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`volar > basic.ts 1`] = `
"import { t as __v_try } from 'tc39-try';
import { Result } from 'try'
function fn() {
  return 10
}
async function fn2() {
  await 10
  return 42
}

export const x: Result<number> = __v_try(() => (fn()))
export const x2: Result<number> = await __v_try(() => (fn2()))"
`;
